src/network/mining_pool.rs

```diff
@@ -0,0 +1,324 @@
+                                warn!("Could not store coinbase record {}", error);
+                            }
+                        });
+
+                        // Broadcast the next block.
+                        let request = LedgerRequest::UnconfirmedBlock(self.local_ip, block, self.prover_router.clone());
+                        if let Err(error) = self.ledger_router.send(request).await {
+                            warn!("Failed to broadcast mined block - {}", error);
+                        }
+                    }
+                } else {
+                    warn!("[ProposedBlock] No current template exists");
+                }
+            }
+            MiningPoolRequest::BlockHeightClear(block_height) => {
+                // Remove the shares for the given block height.
+                if let Err(error) = self.state.remove_shares(block_height) {
+                    warn!("[BlockHeightClear] {}", error);
+                }
+            }
+            MiningPoolRequest::GetCurrentBlockTemplate(peer_ip, address) => {
+                if let Some(current_template) = &*self.current_template.read().await {
+                    // Ensure this miner exists in the info list first, so we can get their share
+                    // difficulty.
+                    if let None = self.miner_info.read().await.get(&address) {
+                        // If not, let's add them.
+                        let share_difficulty = current_template.difficulty_target.saturating_mul(50);
+                        self.miner_info
+                            .write()
+                            .await
+                            .insert(address, (chrono::Utc::now().timestamp(), share_difficulty, 0));
+                    }
+
+                    let share_difficulty = self
+                        .miner_info
+                        .read()
+                        .await
+                        .get(&address)
+                        .expect("miner should exist in miner_info")
+                        .1;
+                    if let Err(error) = self
+                        .peers_router
+                        .send(PeersRequest::MessageSend(
+                            peer_ip,
+                            Message::BlockTemplate(share_difficulty, Data::Object(current_template.clone())),
+                        ))
+                        .await
+                    {
+                        warn!("[ProposedBlock] {}", error);
+                    }
+                } else {
+                    warn!("[ProposedBlock] No current block template exists");
+                }
+            }
+        }
+    }
+
+    async fn set_block_template(&self, recipient: Address<N>) -> Result<()> {
+        let unconfirmed_transactions = self.memory_pool.read().await.transactions();
+        let mut current_template = self.current_template.write().await;
```

Comment:
is this write lock needed here already, i.e. should `prepare_block_template` happen with its contents frozen? otherwise it should only be acquired below it for update purposes